hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -221,48 +218,79 @@ public AzureIngressHandler getIngressHandler() {
+   *
+   * @return The initialized or existing Azure ingress handler.
+   * @throws IOException If an I/O error occurs during handler creation or data processing.
    */
   private AzureIngressHandler createIngressHandler(AbfsServiceType serviceType,
       DataBlocks.BlockFactory blockFactory,
       int bufferSize, boolean isSwitch, AzureBlockManager blockManager) throws IOException {
-    lock.lock();
-    try {
-      this.client = clientHandler.getClient(serviceType);
-      if (switchCompleted && ingressHandler != null) {
-        return ingressHandler; // Return the existing ingress handler
-      }
-      if (isDFSToBlobFallbackEnabled
-          && serviceTypeAtInit == AbfsServiceType.BLOB) {
-        throw new InvalidConfigurationValueException(
-            "The ingress service type must be configured as DFS");
-      }
-      if (isDFSToBlobFallbackEnabled && !isSwitch) {
-        ingressHandler = new AzureDfsToBlobIngressFallbackHandler(this,
-            blockFactory,
-            bufferSize, eTag, clientHandler);
-      } else if (serviceType == AbfsServiceType.BLOB) {
-        ingressHandler = new AzureBlobIngressHandler(this, blockFactory,
-            bufferSize, eTag, clientHandler, blockManager);
-      } else {
-        ingressHandler = new AzureDFSIngressHandler(this, blockFactory,
-            bufferSize, eTag, clientHandler);
-      }
-      if (isSwitch) {
-        switchCompleted = true;
-      }
+    // If the ingressHandler is already initialized and switch is complete, no need to acquire a lock
+    if (ingressHandler != null && switchCompleted) {
       return ingressHandler;
-    } finally {
-      lock.unlock();
+    }
+    // If ingressHandler is not null but the switch is incomplete, lock to safely modify
+    if (ingressHandler != null) {
+      lock.lock();
+      try {
+        // Double-check the condition after acquiring the lock
+        if (ingressHandler != null && switchCompleted) {
+          return ingressHandler; // Return the handler if it's already initialized
+        }
+        // If switch is incomplete, proceed with creating the appropriate handler
+        return createNewHandler(serviceType, blockFactory, bufferSize, isSwitch, blockManager);
+      } finally {
+        lock.unlock();
+      }
+    } else {
+      // ingressHandler is null, no lock is needed, safely initialize it outside the lock
+      return createNewHandler(serviceType, blockFactory, bufferSize, isSwitch, blockManager);
     }
   }
 
+  // Helper method to create a new handler, used in both scenarios (locked and unlocked)
+  private AzureIngressHandler createNewHandler(AbfsServiceType serviceType,
+      DataBlocks.BlockFactory blockFactory,
+      int bufferSize,
+      boolean isSwitch,
+      AzureBlockManager blockManager) throws IOException {
+    this.client = clientHandler.getClient(serviceType);
+    if (isDFSToBlobFallbackEnabled
+        && serviceTypeAtInit == AbfsServiceType.BLOB) {
```

Comment:
Instead of having equals to check with BLOB, should we have not equal to check with DFS? 
`serviceTypeAtInit != AbfsServiceType.DFS`

Reply:
What difference would that make ?

Reply:
taken